### ISSUE TITLE
Helix test runner does not optimally partition tests between work items

### ISSUE THREAD

**Issue by dibarbet** [dotnet/roslyn#62036]
Was pulling data from kusto on what work items take the longest and potential for consolidation and noticed that we have a high variance in how long each work item takes due to issues in partitioning the tests between work items.

For example, `Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests` has the longest average time, simply because it runs all 7000 tests in one item instead of split up into groups of ~500 tests - https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-roslyn-refs-pull-62011-merge-c8a696fc6d2e496eaf/Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests.dll/1/console.e6b822bc.log?helixlogtype=result

And `Microsoft.CodeAnalysis.ExternalAccess.Razor.UnitTests` waits 20min in queue on average to run for only 10seconds because there is only 1 test in the dll
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-roslyn-refs-pull-62011-merge-c8a696fc6d2e496eaf/Microsoft.CodeAnalysis.ExternalAccess.Razor.UnitTests.dll/1/console.7ff23337.log?helixlogtype=result

|WorkItemName|Avg Duration|Avg Time in Queue|Max Duration|
|-----------------|---------------|---------------------|---------------|
|Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests|00:09:22.|00:20:33.|00:24:28.|
|Microsoft.CodeAnalysis.VisualBasic.CommandLine.UnitTests|00:02:00.|00:20:34.|00:06:52.|
|Microsoft.CodeAnalysis.EditorFeatures2.UnitTests|00:01:57.|00:19:20.|00:08:16.|
|Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests|00:01:46.|00:16:00.|00:08:13.|
|Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests|00:01:44.|00:14:20.|00:13:16.|
|Microsoft.CodeAnalysis.CSharp.Emit2.UnitTests|00:01:44.|00:16:27.|00:15:10.|
|Microsoft.CodeAnalysis.CSharp.Symbol.UnitTests|00:01:22.|00:15:53.|00:11:45.|
|Microsoft.CodeAnalysis.CSharp.Emit.UnitTests|00:01:19.|00:15:22.|00:10:51.|
|Microsoft.CodeAnalysis.LanguageServer.Protocol.UnitTests|00:01:16.|00:20:29.|00:04:03.|
|Microsoft.CodeAnalysis.VisualBasic.Emit.UnitTests|00:01:16.|00:14:45.|00:10:21.|
|Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTests|00:01:15.|00:19:33.|00:06:44.|
|Microsoft.CodeAnalysis.EditorFeatures.UnitTests|00:01:07.|00:20:13.|00:04:21.|
|Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests|00:01:05.|00:18:17.|00:06:28.|
|Microsoft.CodeAnalysis.Rebuild.UnitTests|00:01:01.|00:17:06.|00:04:36.|
|Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests|00:00:58.|00:17:05.|00:05:30.|
|Microsoft.CodeAnalysis.CSharp.EditorFeatures2.UnitTests|00:00:57.|00:18:59.|00:06:53.|
|Microsoft.CodeAnalysis.VisualBasic.Symbol.UnitTests|00:00:56.|00:16:11.|00:08:22.|
|Microsoft.CodeAnalysis.VisualBasic.Semantic.UnitTests|00:00:56.|00:15:01.|00:07:30.|
|Microsoft.CodeAnalysis.VisualBasic.CodeStyle.UnitTests|00:00:53.|00:20:16.|00:04:46.|
|Microsoft.CodeAnalysis.UnitTests|00:00:51.|00:16:59.|00:04:47.|
|Microsoft.VisualStudio.LanguageServices.UnitTests|00:00:50.|00:20:04.|00:04:56.|
|Microsoft.CodeAnalysis.CSharp.IOperation.UnitTests|00:00:44.|00:15:43.|00:11:03.|
|Microsoft.CodeAnalysis.Workspaces.UnitTests|00:00:37.|00:19:37.|00:04:07.|
|Microsoft.CodeAnalysis.CSharp.Scripting.Desktop.UnitTests|00:00:36.|00:20:42.|00:02:33.|
|Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests|00:00:34.|00:16:59.|00:03:30.|
|Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.ExpressionCompiler.UnitTests|00:00:33.|00:20:26.|00:05:28.|
|Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.UnitTests|00:00:31.|00:20:41.|00:05:38.|
|Microsoft.CodeAnalysis.CSharp.Syntax.UnitTests|00:00:30.|00:16:38.|00:06:10.|
|Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator.ResultProvider.UnitTests|00:00:30.|00:20:31.|00:03:41.|
|Microsoft.CodeAnalysis.CSharp.Workspaces.UnitTests|00:00:28.|00:19:45.|00:06:19.|
|Microsoft.CodeAnalysis.VisualBasic.Workspaces.UnitTests|00:00:28.|00:19:48.|00:03:46.|
|Microsoft.CodeAnalysis.VisualBasic.Syntax.UnitTests|00:00:27.|00:16:50.|00:05:26.|
|Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator.ExpressionCompiler.UnitTests|00:00:26.|00:20:25.|00:03:28.|
|Microsoft.CodeAnalysis.Remote.ServiceHub.UnitTests|00:00:24.|00:18:32.|00:03:51.|
|Microsoft.CodeAnalysis.CSharp.WinRT.UnitTests|00:00:23.|00:20:24.|00:03:37.|
|Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.ResultProvider.UnitTests|00:00:20.|00:20:36.|00:04:03.|
|Microsoft.CodeAnalysis.VisualBasic.Scripting.UnitTests|00:00:18.|00:20:43.|00:02:15.|
|Microsoft.CodeAnalysis.Workspaces.MSBuild.UnitTests|00:00:17.|00:19:48.|00:05:26.|
|Microsoft.CodeAnalysis.ExpressionEvaluator.FunctionResolver.UnitTests|00:00:17.|00:20:37.|00:03:04.|
|Microsoft.VisualStudio.LanguageServices.LiveShare.UnitTests|00:00:16.|00:20:45.|00:03:50.|
|Microsoft.CodeAnalysis.Scripting.UnitTests|00:00:14.|00:17:05.|00:03:52.|
|Microsoft.CodeAnalysis.ExternalAccess.FSharp.UnitTests|00:00:14.|00:20:45.|00:01:59.|
|Microsoft.CodeAnalysis.Workspaces.Desktop.UnitTests|00:00:12.|00:26:00.|00:01:09.|
|Microsoft.CodeAnalysis.Scripting.Desktop.UnitTests|00:00:11.|00:20:45.|00:01:53.|
|Microsoft.CodeAnalysis.VisualBasic.Scripting.Desktop.UnitTests|00:00:10.|00:20:48.|00:01:52.|
|Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.UnitTests|00:00:10.|00:20:49.|00:02:49.|
|Microsoft.CodeAnalysis.ExternalAccess.Razor.UnitTests|00:00:10.|00:20:43.|00:02:09.|

**Comment by dibarbet** [comment]
My current plan of action is to fix the partition logic to always ensure we always have a consistent number of tests to run regardless of dll (e.g. grouping test dlls until we have a large number to run together and fixing the partition logic for `Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests`

This should hopefully allow us to get to a roughly consistent work item runtime that is less than the current max runtime and use less machines in total.

**Comment by RikkiGibson** [comment]
Really great catch David.

**Comment by dibarbet** [comment]
The main problem with `Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests` that is causing all the tests to run in one work item is the usage of combinatorial tests.  Currently RunTests looks at annotated test methods to figure out how many tests are in a particular class.  However, `AbstractPersistentStorageTests` defines combinatorial tests that run 20 iterations along with other combinatorial parameters that explode ~20 test methods into ~3.5k actual tests to run.

https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests/PersistentStorage/AbstractPersistentStorageTests.cs,132

I do not have a good solution for this - it requires some amount of introspection on the methods themselves to calculate the combinations (e.g. enum (4) * bool (2) * iterations (20)).

**Comment by dibarbet** [comment]
Jon pointed me to `dotnet test --list-tests` which we should be able to use instead of the current reflection based method to figure out how many tests we have in a dll

**Comment by RikkiGibson** [comment]
Do we need to spawn 'dotnet test' subprocesses to get that info? Not possible to use a library method or anything?

FWIW, I am hoping to investigate a larger task later this year to try and implement an XUnit.TestFramework which partitions and runs tests in subprocesses. Don't know yet whether it's possible but hoping very much to be able to 'dotnet test path/to/assembly' and get similar parallelism as '.\Test.cmd'.

**Comment by dibarbet** [comment]
> Do we need to spawn 'dotnet test' subprocesses to get that info? Not possible to use a library method or anything?

It is possible there is one, but I haven't seen anything obvious other than dotnet test.

> FWIW, I am hoping to investigate a larger task later this year to try and implement an XUnit.TestFramework which partitions and runs tests in subprocesses. Don't know yet whether it's possible but hoping very much to be able to 'dotnet test path/to/assembly' and get similar parallelism as '.\Test.cmd'.

That sounds great actually!  For now in the short term I think I can get RunTests to work a little better, but am all for a better approach.  It would be great to be able to just run dotnet test.

**Comment by RikkiGibson** [comment]
We discussed internally that it might be worthwhile to try and have the test job pull in the results from the previous run of the corresponding test job.

e.g. https://dev.azure.com/dnceng/public/_apis/test/Runs/49077962/results?api-version=6.0

using the REST API documented here: [Results - List - REST API (Azure DevOps Test) | Microsoft Docs](https://docs.microsoft.com/en-us/rest/api/azure/devops/test/results/list?view=azure-devops-rest-6.0)

Basically, instead of just counting up test methods to make a partition, we could sum up execution times from the previous run. I think our granularity is still at the class level, so we might still end up limited by the fact that some classes may take a long time to run.

**Comment by dibarbet** [comment]
> I think our granularity is still at the class level, so we might still end up limited by the fact that some classes may take a long time to run.

The ADO api appears to output some test name (though I'll have to check what it does for theory tests).  With that plus using [response files](https://docs.microsoft.com/en-us/dotnet/standard/commandline/syntax#response-files) to get around the command line length limit we might be able to actually filter by individual test cases

**Comment by dibarbet** [comment]
Plan to pick back up work on this on Monday

**Comment by dibarbet** [comment]
Update - I'm able to pretty successfully get historical runtimes.  However partitioning just based on type name is not an option, as some types take longer than 5 minutes to run, for example
- Work Item (1 types, runtime 00:08:12.3060000)
    - Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests.dll with 1 types, runtime 00:08:12.3060000

I need to update the SRM code to correctly identify possible test methods (last roadblock I hit was test methods in abstract classes that apply to the concrete impl that has no tests).

Or alternatively switch back to actual test discovery on the dlls, but that is unfortunately rather slow, even when using the test platform API directly https://www.nuget.org/packages/Microsoft.TestPlatform.TranslationLayer

**Comment by dibarbet** [comment]
System.Reflection.Metadata is extremely brittle for this.  Many of the base types that have the actual test methods are in different assemblies which I don't think I can easily get to with SRM (especially since we've minimized the payload and don't even have all the assemblies next to the test dll).

I've tried a few things like just considering the entire type if we have base types, but all of this adds a ton of complexity and tons of chances to accidentally drop tests.

My next path of investigation is to run actual discovery (which as before can take multiple minutes), but then attempt to partition into work items that run in less than the time we took to run discovery (so overall still run under 5 minutes).  I think this may be doable even while using _less_ partitions than we do today.

**Comment by RikkiGibson** [comment]
This is probably a silly idea, but what if we added a source generator which wrote out the partitions for each test project during the build?

**Comment by dibarbet** [comment]
> This is probably a silly idea, but what if we added a source generator which wrote out the partitions for each test project during the build?

Definitely a possibility.  However it still means we're almost re-implementing test discovery which we could definitely get wrong.  This is what I'll do
1.  See if I can run test discovery fast enough and make enough partitions that discovery + test execution takes 5 minutes, with fewer or the same number of partitions that we use today.
2.  If 1) doesn't work - I'll start going down the source generator route.  I think it is absolutely viable, but I'd rather discover what actually gets run if possible.

**Comment by dibarbet** [comment]
If we do the source generator route - I think we probably don't want to actually compute the partitions in the generator - that would require the generator to query ADO for the test history.  Instead we'd have it write out the tests per assembly to a file and then handle the partitioning during the build.

**Comment by dibarbet** [comment]
@RikkiGibson the test `Microsoft.CodeAnalysis.CSharp.UnitTests.OverloadResolutionPerfTests.NestedLambdas_01` seems to fairly consistently take longer than 5 minutes to run.  If we have a single test that takes that long we definitely can't get our CI times below that.  Thoughts on if this test can be modified or split up into multiple shorter runs?

![image](https://user-images.githubusercontent.com/5749229/180591701-982485ee-3eec-4f1c-8656-872d71bd9207.png)
https://dev.azure.com/dnceng/public/_build/results?buildId=1898809&view=ms.vss-test-web.build-test-results-tab&runId=49434452&resultId=223986&paneView=history
